lighthouse-core/runner.js

```diff
@@ -56,83 +59,68 @@ class Runner {
     // canonicalize URL with any trailing slashes neccessary
     opts.url = parsedURL.href;
 
-    // Check that there are passes & audits...
-    const validPassesAndAudits = config.passes && config.audits;
-
-    // ... or that there are artifacts & audits.
-    const validArtifactsAndAudits = config.artifacts && config.audits;
-
     // Make a run, which can be .then()'d with whatever needs to run (based on the config).
     let run = Promise.resolve();
 
-    // If there are passes run the GatherRunner and gather the artifacts. If not, we will need
-    // to check that there are artifacts specified in the config, and throw if not.
-    if (validPassesAndAudits || validArtifactsAndAudits) {
-      if (validPassesAndAudits) {
-        // Set up the driver and run gatherers.
-        opts.driver = opts.driverMock || new Driver(connection);
-        run = run.then(_ => GatherRunner.run(config.passes, opts));
-      } else if (validArtifactsAndAudits) {
-        run = run.then(_ => config.artifacts);
-      }
-
-      run = run.then(artifacts => {
-        // Bring in lighthouseRunWarnings from gathering stage.
-        if (artifacts.LighthouseRunWarnings) {
-          lighthouseRunWarnings.push(...artifacts.LighthouseRunWarnings);
-        }
-
-        // And add computed artifacts.
-        return Object.assign({}, artifacts, Runner.instantiateComputedArtifacts());
-      });
+    // User can run -G solo, -A solo, or -GA together
+    const shouldGatherAndQuit = opts.flags.gatherMode && !opts.flags.auditMode;
+    const shouldOnlyAudit = opts.flags.auditMode && !opts.flags.gatherMode;
+    const shouldDefaultRunButSaveArtifacts = opts.flags.auditMode && opts.flags.gatherMode;
+    const shouldDoTypicalRun = !opts.flags.gatherMode && !opts.flags.auditMode;
 
-      // Basic check that the traces (gathered or loaded) are valid.
-      run = run.then(artifacts => {
-        for (const passName of Object.keys(artifacts.traces || {})) {
-          const trace = artifacts.traces[passName];
-          if (!Array.isArray(trace.traceEvents)) {
-            throw new Error(passName + ' trace was invalid. `traceEvents` was not an array.');
-          }
-        }
+    const shouldSaveArtifactsToDisk = shouldGatherAndQuit || shouldDefaultRunButSaveArtifacts;
+    const shouldLoadArtifactsFromDisk = shouldOnlyAudit;
+    const shouldGatherFromBrowser = config.passes && !config.artifacts &&
+        (shouldGatherAndQuit || shouldDefaultRunButSaveArtifacts || shouldDoTypicalRun);
 
-        return artifacts;
+    if (shouldLoadArtifactsFromDisk) {
+      config.removePasses();
+      run = run.then(_ => {
+        return assetSaver.loadArtifacts(basePath).then(artifacts => config._artifacts = artifacts);
```

Comment:
could we collect config's artifacts into a local variable and just manipulate that instead of relying on them being set on the config down below?